OutOfSchool/OutOfSchool.WebApi.Tests/Services/ParentServiceTests.cs

```diff
@@ -48,9 +53,170 @@ public void SetUp()
+
+        // Act & Assert
+        Assert.ThrowsAsync<InvalidOperationException>(async () => await parentService.Create(new ParentCreateDto()).ConfigureAwait(false));
+
+        loggerMock.Verify(
+            logger => logger.Log(
+                It.Is<LogLevel>(logLevel => logLevel == LogLevel.Error),
+                It.Is<EventId>(eventId => eventId.Id == 0),
+                It.IsAny<It.IsAnyType>(),
+                It.IsAny<Exception>(),
+                It.IsAny<Func<It.IsAnyType, Exception, string>>()),
+            Times.Once);
+    }
+
+    [Test]
+    public void Create_WhenUserNotExists_ShouldLogErrorAndThrowInvalidOperationException()
+    {
+        // Arrange
+        currentUserServiceMock.SetupGet(s => s.UserId).Returns("userId");
+
+        userRepositoryMock
+            .Setup(r => r.GetById("userId"))
+            .ReturnsAsync(null as User);
+
+        // Act & Assert
+        Assert.ThrowsAsync<InvalidOperationException>(async () => await parentService.Create(new ParentCreateDto()).ConfigureAwait(false));
+
+        loggerMock.Verify(
+            logger => logger.Log(
+                It.Is<LogLevel>(logLevel => logLevel == LogLevel.Error),
+                It.Is<EventId>(eventId => eventId.Id == 0),
+                It.IsAny<It.IsAnyType>(),
+                It.IsAny<Exception>(),
+                It.IsAny<Func<It.IsAnyType, Exception, string>>()),
+            Times.Once);
+    }
+
+    [Test]
+    public void Create_WhenParentExists_ShouldLogErrorAndThrowInvalidOperationException()
+    {
+        // Arrange
+        currentUserServiceMock.SetupGet(s => s.UserId).Returns("userId");
+
+        parentRepositoryMock
+            .Setup(r => r.Any(It.IsAny<Expression<Func<Parent, bool>>>()))
+            .ReturnsAsync(true);
+
+        // Act & Assert
+        Assert.ThrowsAsync<InvalidOperationException>(async () => await parentService.Create(new ParentCreateDto()).ConfigureAwait(false));
+
+        loggerMock.Verify(
+            logger => logger.Log(
+                It.Is<LogLevel>(logLevel => logLevel == LogLevel.Error),
+                It.Is<EventId>(eventId => eventId.Id == 0),
+                It.IsAny<It.IsAnyType>(),
+                It.IsAny<Exception>(),
+                It.IsAny<Func<It.IsAnyType, Exception, string>>()),
+            Times.Once);
+    }
+
+    [Test]
+    public async Task Create_WhenUserExists_ShouldSetUserIsRegistered()
+    {
+        // Arrange
+        var user = UserGenerator.Generate();
+
+        currentUserServiceMock.SetupGet(s => s.UserId).Returns(user.Id);
+
+        userRepositoryMock
+            .Setup(r => r.GetById(user.Id))
+            .ReturnsAsync(user);
+
+        parentRepositoryMock
+            .Setup(r => r.Create(It.IsAny<Parent>()))
+            .ReturnsAsync(new Parent());
+
+        parentRepositoryMock
+            .SetupGet(r => r.UnitOfWork)
+            .Returns(new Mock<IUnitOfWork>().Object);
```

Comment:
(Test logic) There are few issues with selected code, originated from weird Sonar coverage calculation algorithm, violation of Demeter's law done in design of our repositories, and lack of testing. Please consider (in this and other tests) applying changes sugested below:

1. replace in `Arrange` section
```
        parentRepositoryMock
            .SetupGet(r => r.UnitOfWork)
            .Returns(new Mock<IUnitOfWork>().Object);
``` 
by
``` 
        parentRepositoryMock.Setup(r => r.UnitOfWork.CompleteAsync()) .Returns(Task.FromResult(1));`
``` 

2. and add in `Assert` section
``` 
        parentRepositoryMock.Verify(r => r.UnitOfWork.CompleteAsync(), Times.Once);`
``` 

---

More explanations

As I understand the selected code is used to support next line of code in `ParentService.Create()`
`        await repositoryParent.UnitOfWork.CompleteAsync();`

Imagine that `UnitOfWork` also implements `RollbackAsync()` method (now it doesn't and it's really strange, but that's other point). If code of `ParentService.Create()` by mistake calls `RollbackAsync()` instead of currently used `CompleteAsync()` your test will not fail. But it should as data are not created! Thus testing logic has gap.

Unfortunately Sonar thinks that that specific line is covered, thus you would never be forced to close this gap.

And yes, as without selected mocking code test fails with NRE, it's enough to provide mocks for your mocks, but this is usually a sign of a smell in design - viloation of Demeter's law.

Demeter's law states (among other statements) - "Only talk to your immediate friends". And since the time when we'd introduced public `UnitOfWork` to our repository API, and started to use it via call chain like `repo.UnitOfWork.Complete...()` (in total there are 11 places in our code) we are violating this statement. As a result tests are also "broken" - extra dependency is introduced that increases coupling and thus makes tests more fragile.

So, origin of the problem is design of repository API. But, as redesign is out of this PR scope, please at least make test more stable and "covering" as suggested (btw, this is already used in other tests).

Reply:
> Demeter's law states (among other statements) - "Only talk to your immediate friends". And since the time when we'd introduced public `UnitOfWork` to our repository API, and started to use it via call chain like `repo.UnitOfWork.Complete...()` (in total there are 11 places in our code) we are violating this statement. As a result tests are also "broken" - extra dependency is introduced that increases coupling and thus makes tests more fragile.
> 
> So, origin of the problem is design of repository API. But, as redesign is out of this PR scope, please at least make test more stable and "covering" as suggested (btw, this is already used in other tests).

Thank you for the explanation! This information is really helpful. Although I heard about Demeter’s law, I never encountered a violation of it in practice.